blockchain/internal/spendpruner/pruner_test.go

```diff
@@ -0,0 +1,346 @@
+	}
+
+	fetchHashSpendDependents := func(pruner *SpendJournalPruner, hash *chainhash.Hash) ([]string, bool) {
+		pruner.dependentsMtx.Lock()
+		dependents, ok := pruner.dependents[*hash]
+		pruner.dependentsMtx.Unlock()
+
+		return dependents, ok
+	}
+
+	isRemovedSpendEntry := func(chain *testSpendPurger, hash *chainhash.Hash) bool {
+		chain.mtx.Lock()
+		_, ok := chain.removedSpendEntries[*hash]
+		chain.mtx.Unlock()
+
+		return ok
+	}
+
+	consumerB.needSpendDataErr = nil
+
+	// Ensure adding dependents creates entries for consumers that need
+	// the spend data of the provided block hash.
+	err = pruner.addSpendConsumerDeps(hashA)
+	if err != nil {
+		t.Fatalf("unexpected error adding consumer dependencies: %v", err)
+	}
+
+	// Ensure there is only one dependent entry for hashA, belonging to
+	// consumerA.
+	dependents, ok := fetchHashSpendDependents(pruner, hashA)
+	if !ok {
+		t.Fatalf("expected dependents to have an entry for %s", hashA)
+	}
+
+	if len(dependents) != 1 {
+		t.Fatalf("expected one dependent entry for hashA, got %d",
+			len(dependents))
+	}
+
+	if dependents[0] != consumerA.ID() {
+		t.Fatalf("expected hashA's dependent to be %s, got %s",
+			consumerA.ID(), dependents[0])
+	}
+
+	// Ensure there is no existing dependency for consumerB on hashA.
+	exists := pruner.dependencyExistsInternal(hashA, consumerB.ID())
+	if exists {
+		t.Fatal("unexpected dependency found for consumerB on hashA")
+	}
+
+	// Ensure there is an existing dependency for consumerA on hashA.
+	exists = pruner.dependencyExistsInternal(hashA, consumerA.ID())
+	if !exists {
+		t.Fatal("expected dependency found for consumerA on hashA")
+	}
+
+	// Ensure consumerC and consumerA now have dependencies on hashA.
+	consumerC := &testSpendConsumer{
+		id:            "consumer_c",
+		tip:           &chainhash.Hash{0},
+		needSpendData: true,
+	}
+
+	pruner.AddConsumer(consumerC)
+
+	setConsumerNeedSpendData := func(state bool) {
+		consumerA.needSpendData = state
+		consumerB.needSpendData = state
+		consumerC.needSpendData = state
+	}
+
+	err = pruner.addSpendConsumerDeps(hashA)
+	if err != nil {
+		t.Fatalf("unexpected error adding spend data dependents "+
+			"for hashA: %v", err)
+	}
+
+	exists = pruner.DependencyExists(hashA)
+	if !exists {
+		t.Fatal("expected existing dependencies for hashA")
+	}
+
+	exists = pruner.dependencyExistsInternal(hashA, consumerC.ID())
+	if !exists {
+		t.Fatal("expected dependency found for consumerC on hashA")
+	}
+
+	exists = pruner.dependencyExistsInternal(hashA, consumerA.ID())
+	if !exists {
+		t.Fatal("expected dependency found for consumerA on hashA")
+	}
+
+	// Ensure there are now two dependent entries for hashA.
+	dependents, _ = fetchHashSpendDependents(pruner, hashA)
+	if len(dependents) != 2 {
+		t.Fatalf("expected two dependent entry for hashA, got %d",
+			len(dependents))
+	}
+
+	// Ensure removing a non-existent dependency does nothing.
+	pruner.RemoveSpendConsumerDependency(hashA, "consumer_x")
```

Comment:
Should check for an error where `RemoveSpendConsumerDependency` is called (here and below) since it was updated to return an error.